### PR TITLE
Return all users lists with mentors

### PR DESCRIPTION
**Pull request by MuthomiMate**
It creates an endpoint to return all users lists with mentors

It closes [#99 ](https://github.com/Coding-Coach/find-a-mentor-api/issues/99)

### CODE REVIEW

File: src/modules/lists/lists.controller.ts [hunk 1]
```diff
@@ -45,4 +45,31 @@ export class ListsController {
...
+    const lists: List[] = await this.listsService.findByUserId(userId);
+
+    // Only current user and admins can view both private and public lists for a user
+    if (current._id.equals(user._id) || current.roles.includes(Role.ADMIN)) {
+      visibleList = lists;
+    } else {
+      if (Array.isArray(lists)) {
+        visibleList = lists.filter(list => list.public === true);
+      }
+    }
```
Comment (crysfel) [inline src/modules/lists/lists.controller.ts]: Let's leverage the filtering to the database, by using filters like here: https://github.com/Coding-Coach/find-a-mentor-api/blob/master/src/modules/common/mentors.service.ts#L45

```
const lists: List[] = await this.listsService.find({ _id: '12345', public: true|false });
```

File: src/modules/lists/lists.service.ts [hunk 2]
```diff
@@ -13,4 +14,12 @@ export class ListsService {
     const list = new this.listModel(data);
     return await list.save();
   }
+
+  async findByUserId(_id: string): Promise<List[]> {
+    if (isObjectId(_id)) {
+      return await this.listModel.find({user: {_id}}).exec();
```
Comment (crysfel) [inline src/modules/lists/lists.service.ts]: In here let's also filter by `public`, the controller should decide the value of this filter based on the user roles.

File: src/modules/lists/lists.controller.ts [hunk 3]
```diff
@@ -45,4 +45,30 @@ export class ListsController {
...
+
+    // check if user exists
+    if (!user) {
+      throw new BadRequestException('User not found');
+    }
+    let lists: List[];
+
+    // Only current user and admins can view both private and public lists for a user
+    if (current._id.equals(user._id) || current.roles.includes(Role.ADMIN)) {
+      lists = await this.listsService.findByUserId({ _id: userId, public: [true, false]});
```
Comment (crysfel) [inline src/modules/lists/lists.controller.ts]: If you don't send a `public` value here, then the service should ignore this filter and should return both values.

```suggestion
lists = await this.listsService.find({ _id: userId });
```

File: src/modules/lists/lists.controller.ts [hunk 4]
```diff
@@ -45,4 +45,30 @@ export class ListsController {
...
+    }
+    else{
+      lists = await this.listsService.findByUserId({ _id: userId, public: [true] });
```
Comment (crysfel) [inline src/modules/lists/lists.controller.ts]: If we want to filter by `public` (regardless of `true` or `false` value) we should send the filter in this case

```suggestion
lists = await this.listsService.find({ _id: userId, public: true });
```

File: src/modules/lists/lists.service.ts [hunk 5]
```diff
@@ -13,4 +14,11 @@ export class ListsService {
     const list = new this.listModel(data);
     return await list.save();
   }
+
+  async findByUserId({_id, public: isPublic }): Promise<List[]> {
```
Comment (crysfel) [inline src/modules/lists/lists.service.ts]: Given that we can use this method to filter with by of the fields (maybe later we can filter by name or something else), let's rename it to just `find`.

```suggestion
async find(filters): Promise<List[]> {
```

File: src/modules/lists/lists.service.ts [hunk 6]
```diff
@@ -13,4 +14,11 @@ export class ListsService {
...
+    if (isObjectId(_id)) {
+      return await this.listModel.find({user: {_id}, public: {$in: isPublic}}).exec();
```
Comment (crysfel) [inline src/modules/lists/lists.service.ts]: You should build the filters dynamically based on the params received:

```
const filters = {};

if (filters.userId) {
  filters.user = { _id: userId };
}

if (filters.public !== undefined) {
  filters.public = public;
}

// TODO: Add more filters here later on (as we need them)

return await this.listModel.find(filters).exec();
```

**Comment by MuthomiMate** [comment]
@crysfel I have implemented the requested changes

**Pull request merged**